docs/RFCS/20190619_constraint_conformance_report.md

```diff
@@ -0,0 +1,319 @@
+  for under-diversification - i.e. we'll report a range as under-diversified
+  even if no replica can be moved at the moment because of space constraints. 
+  Note that when the under-diversification is due to over-replication (i.e. the
+  only violation is that two replicas are in the same locality), that’d be
+  accounted for in the over-replication stats, without also being counted here.
+
+
+Besides raw numbers, this report will also include velocity information (the
+`change_rate` col): how did each number change in the last *x* minutes. This can
+be `NULL` if there's no previous data on a particular violation.
+
+## AdminUI conformance report
+
+We’ll add a page in the AdminUI showing the contents of the virtual table. In
+the AdminUI we can do fancier things like allowing one to drill down into a zone
+- expand a zone into its constituent tables.
+
+## New jobs
+
+We’re also going to have some data-moving operations roughly represented as
+jobs: repartitioning and altering zone config properties. The idea here is to be
+able to track the progress of an operation that was explicitly triggered through
+a SQL command as opposed to observing the state of the cluster as a whole.
+For example, when a table is repartitioned, we’ll look at how much data of that
+table is now placed incorrectly (i.e. it violates a constraint) and we’ll
+consider the job complete once all the table’s data is conformant with all the
+constraints. Note that the initial partitioning of a table does not create any
+data movement (as all the partitions inherit the same zone config) and so we
+won’t create a job for it.
+
+Altering a zone config (for a table, partition or index) will similarly cause a
+job to be created. The job will be considered completed once all the ranges
+corresponding to the table/index/partition are in conformance with all the
+constraints. That is, the first time when all ranges in the affected zone are
+found to be in compliance, we’ll declare success.
+
+Note that the association of range movement with a job, and thus the computation
+of the job’s progress, is fairly loose. The exact replication effects being
+enacted by a particular partition or zone change will be immaterial to the
+progress status of the respective job; the only thing will be considered for
+computing the progress is the conformance of the ranges affected by the change
+with all the constraints (not just the constraints being modified) - including
+pre-existing constraints for the respective zone and constraints inherited from
+parent zones. Doing something more exact seems hard. However this can lead to
+funny effects like a progress going backwards if the conformance worsens (i.e.
+number of con-conformant ranges increases for whatever reason) while a job is
+“running”.
+
+The updating of the progress report of these jobs will be done through the
+periodic background process that also updates the cluster conformance report.
+The jobs are not cancelable.
```

Comment:
What happens if a user does a conflicting job before the first job finishes? I assume we would want to show the first job as canceled/superseded by the new job. How would we detect this? How would we present it in the ui?

Reply:
specified that a job is marked as completed if the operation is superseded. I don't think the UI needs to have anything in particular to reflect these. My plan is to show the old job as completed and a new job as in-progress.

Reply:
cool!